File: Asteroid Collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        int n=asteroids.length;
        for(int i=0;i<n;i++)
        {
            if(asteroids[i]<=0)
            {
                while(!stack.isEmpty() && stack.peek ()> 0 && stack.peek()<asteroids[i]*-1)
                    stack.pop();
                
                if(!stack.isEmpty() && stack.peek()==asteroids[i]*-1)
                    stack.pop();
                
                else if(!stack.isEmpty() && stack.peek() >= -1*asteroids[i])
                    continue;
                else
                    stack.push (asteroids[i]);
            }
        else
            stack.push(asteroids[i]);
            
            
        }
        
        n=stack.size();
        int arr[]=new int[n];
        
        for(int i=n-1;i>=0;i--)
        arr[i]=stack.pop();
        return arr;
        
    }
}

File: Asteroid Collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) 
    {
        stack<int>s;

        for(int asteroid : asteroids)
        {
             int flag = 1;
             while(!s.empty() && s.top() > 0 && asteroid < 0)
             {
                 if( abs(asteroid) > s.top())
                 {
                     s.pop();
                     continue;
                 }

                 if( abs(asteroid) == s.top())
                 {
                     s.pop();
                 }

                 flag = 0;
                 break;
             }

             if(flag)
             {
                 s.push( asteroid );
             } 
        }

        vector<int>ans;
        while(!s.empty())
        {
            ans.push_back(s.top());
            s.pop();
        }

        reverse(ans.begin(), ans.end());
        return ans;
    }
};
